src/live/emitters.js

```diff
@@ -0,0 +1,15 @@
+const { io } = require("../expressServer");
+
+function sendOrderCancelledEvent(ordersToUpdate) {
+  console.log("ordersToUpdate: ", ordersToUpdate);
+  
+  for (const userData of ordersToUpdate) {
+    const userId = userData._id.toString();
+    const numOrders = userData.orders.length;
+    const orderNoun = numOrders === 1 ? 'order' : 'orders';
```

Comment:
In Order.js, we loop through the aggregated \`ordersToUpdate\` and pass in just the object for a given user's orders to this emitter.js function. So we expect this function to receive an object that looks like:

```javascript  
{   
 _id: <user id>,  
 orders: <list of orders that have been cancelled for this user for being abandoned too long>  
}  
```

So there is nothing to iterate/loop over here, we can just access these variables directly from the object passed in:

```suggestion  
function sendOrderCancelledEvent(userData) { 

  console.log("ordersToUpdate: ", userData);  
   
  const userId = userData.\_id.toString();  
  const numOrders = userData.orders.length;  
  const orderNoun = numOrders === 1 ? 'order' : 'orders';  
```

Reply:
<img width="650" alt="screen1" src="https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/95896159/8cee0b61-6046-4cc9-8b9e-5e154b81c648">

Reply:
<img width="636" alt="screen5" src="https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/95896159/49e4fcf1-fa6e-47eb-8164-2812e8cf8e5a">

Reply:
Thank you very much! I've sent screenshots of what I'm getting in the browser and terminal. I don't quite understand how to test this code in Postman. I'm establishing a connection with the socket and Listening to
order:cancelled.

Reply:
Sorry for the delay in responding to this question:

You'll want to follow the same steps as what's happening to the user in the browser on the `orderNotifications.js` page:

1.  Connect to the socket on `http://localhost:8000` :white\_check\_mark: 
2.  Emit a message to the `join:user_notifications` channel with the user ID whose notifications you want to receive. This is needed because for the order cancelled notification we only send it to the specific "room" for that user id (since we don't want other users to get notifications about your cancelled orders). I think that part is missing in your screenshot. You'll need to do this in the "Message" tab of the Postman socket request. When you send this message from Postman, then in the backend terminal you should see the following text logged: `User joined room for notifications: <your user id>`
3.  Subscribe to events for **`orders`**`:cancelled` channel. Your screenshot has it mis-spelled with the singular **`order`**`:cancelled` so that may have caused some issues testing it out. 
4.  Create an order (POST request using POSTMAN to the `/api/v1/orders` endpoint) to create a 'pending' order. Allow some time to let the backend convert that 'pending' order to a 'cancelled' one. To speed things up for local testing, I set my expiry time to 15 seconds using the `DEV_ORDER_EXPIRY_DURATION=15000` environment variable in my `.env` file. 

![](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/9094098/d86b8388-fd20-4354-8a6c-ed1717e07edf)

![](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/9094098/2c57b0f9-728b-423d-86c6-77b084c0c067)

Reply:
Its tested for me too, thank you very much!